File: go/0022.generate-parentheses/solution.go
// Created by Olgierd Palasz at 2026/05/06 21:33
// leetgo: dev
// https://leetcode.com/problems/generate-parentheses/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// @lc code=begin

func generateParenthesis(n int) []string {
	result := []string{}
	backtrack(&result, "", 0, 0, n)
	return result
}

func backtrack(result *[]string, current string, open int, close int, max int) {
	if len(current) == 2*max {
		*result = append(*result, current)
		return
	}
	if open < max {
		backtrack(result, current+"(", open+1, close, max)
	}
	if close < open {
		backtrack(result, current+")", open, close+1, max)
	}
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	n := Deserialize[int](ReadLine(stdin))
	ans := generateParenthesis(n)

	fmt.Println("\noutput:", Serialize(ans))
}

File: rust/src/0022.generate-parentheses/solution.rs
// Created by Olgierd Palasz at 2026/08/05 21:05
// leetgo: dev
// https://leetcode.com/problems/generate-parentheses/

use std::unreachable;

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin

impl Solution {
    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        let mut stack: Vec<String> = Vec::new();
        let mut result: Vec<String> = Vec::new();

        Self::generate(&mut stack, &mut result, 0, 0, n as usize);

        result
    }

    fn generate(
        stack: &mut Vec<String>,
        result: &mut Vec<String>,
        open_n: usize,
        closed_n: usize,
        n: usize,
    ) {
        if open_n == n && closed_n == n {
            result.push(stack.concat());
            return;
        }

        if open_n < n {
            stack.push("(".to_string());
            Self::generate(stack, result, open_n + 1, closed_n, n);
            stack.pop();
        }

        if closed_n < open_n {
            stack.push(")".to_string());
            Self::generate(stack, result, open_n, closed_n + 1, n);
            stack.pop();
        }
    }
}

// @lc code=end

fn main() -> Result<()> {
    let n: i32 = deserialize(&read_line()?)?;
    let ans: Vec<String> = Solution::generate_parenthesis(n).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
